File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/chemical/ChemicalVerification.kt
package com.rork.vinetrack.data.chemical

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Where a piece of chemical information came from, ranked by how much weight it
 * may carry.
 *
 * The ordering IS the source hierarchy: an official register outranks a
 * manufacturer label, which outranks an authoritative activity-group
 * classification for identity purposes, and AI/search interpretation sits at the
 * bottom. [AUTHORITATIVE_CLASSIFICATION] is the highest authority for the
 * activity group specifically — that is the one thing FRAC/HRAC/IRAC define.
 */
@Serializable
enum class ChemicalDataSourceKind(val raw: String, val label: String) {
    /** A national regulator's registered-product record: APVMA (AU), ACVM/EPA (NZ). */
    @SerialName("official_register")
    OFFICIAL_REGISTER("official_register", "Official register"),

    /** The registrant's own approved label document. */
    @SerialName("manufacturer_label")
    MANUFACTURER_LABEL("manufacturer_label", "Product label"),

    /** FRAC / HRAC / IRAC classification. Authoritative for activity group and nothing else. */
    @SerialName("authoritative_classification")
    AUTHORITATIVE_CLASSIFICATION("authoritative_classification", "Activity group classification"),

    /** An industry spray guide used to cross-check that a product is used on grapes. */
    @SerialName("viticulture_reference")
    VITICULTURE_REFERENCE("viticulture_reference", "Viticulture reference"),

    /**
     * A language model's reading of search results. Useful for FINDING a
     * candidate; never sufficient on its own to call anything Verified.
     */
    @SerialName("ai_interpretation")
    AI_INTERPRETATION("ai_interpretation", "AI/search interpretation"),

    @SerialName("manual_entry")
    MANUAL_ENTRY("manual_entry", "Manually entered"),

    /** Read out of a pre-Chemical-Intelligence free-text field. */
    @SerialName("legacy_record")
    LEGACY_RECORD("legacy_record", "Existing VineTrack record"),
    ;

    /**
     * Whether this source can, by itself, support a Verified claim.
     *
     * This single property is what stops AI confidence from being laundered into
     * authority.
     */
    val isAuthoritative: Boolean
        get() = when (this) {
            OFFICIAL_REGISTER, MANUFACTURER_LABEL, AUTHORITATIVE_CLASSIFICATION -> true
            VITICULTURE_REFERENCE, AI_INTERPRETATION, MANUAL_ENTRY, LEGACY_RECORD -> false
        }

    /**
     * Whether this source is the record telling us about itself rather than
     * anything outside VineTrack having been consulted.
     *
     * A manually typed value and a value read out of an old free-text column are
     * both statements by the operator's own installation. Neither is proof of
     * anything, which is why a hand-entered registration number cannot make a
     * product's identity authoritative.
     */
    val isSelfReported: Boolean get() = this == MANUAL_ENTRY || this == LEGACY_RECORD

    /** Higher wins when two sources disagree about the same field. */
    val precedence: Int
        get() = when (this) {
            OFFICIAL_REGISTER -> 100
            MANUFACTURER_LABEL -> 90
            AUTHORITATIVE_CLASSIFICATION -> 80
            VITICULTURE_REFERENCE -> 50
            AI_INTERPRETATION -> 20
            MANUAL_ENTRY -> 15
            LEGACY_RECORD -> 10
        }

    companion object {
        /**
         * An unknown source kind from a newer build must never be READ AS
         * authoritative. Falling back to AI interpretation is the safe
         * direction: it can only lower a verification claim, never raise one.
         */
        fun from(raw: String?): ChemicalDataSourceKind? {
            val v = raw?.trim()?.lowercase()?.takeIf { it.isNotEmpty() } ?: return null
            return entries.firstOrNull { it.raw == v } ?: AI_INTERPRETATION
        }
    }
}

/** One cited source behind a chemical's information. */
@Serializable
data class ChemicalDataSource(
    val kind: ChemicalDataSourceKind = ChemicalDataSourceKind.AI_INTERPRETATION,
    /** e.g. `"APVMA PUBCRIS"`, `"FRAC Code List 2025"`. */
    val name: String = "",
    /** URL or document identifier, where one exists. */
    val reference: String? = null,
    @SerialName("retrieved_at") val retrievedAt: String? = null,
) {
    val id: String get() = "${kind.raw}|$name|${reference.orEmpty()}"
}

/** Whether any cited source can support a Verified claim. */
fun List<ChemicalDataSource>.containsAuthoritative(): Boolean = any { it.kind.isAuthoritative }

/** The strongest source cited. */
fun List<ChemicalDataSource>.strongest(): ChemicalDataSource? = maxByOrNull { it.kind.precedence }

/**
 * How much VineTrack trusts a chemical's resistance-critical information.
 *
 * Deliberately separate from "did the lookup return anything". An AI answer is a
 * lead, not a verification: a product only becomes [VERIFIED] when an
 * authoritative source stands behind its identity AND behind every active's
 * activity group.
 */
@Serializable
enum class ChemicalVerificationStatus(val raw: String, val label: String) {
    @SerialName("verified")
    VERIFIED("verified", "Verified"),

    @SerialName("partially_verified")
    PARTIALLY_VERIFIED("partially_verified", "Partially verified"),

    @SerialName("unverified")
    UNVERIFIED("unverified", "Unverified"),

    /**
     * A legacy record that has never been put through the match step. It has
     * data, but nobody has yet confirmed WHICH registered product it is.
     */
    @SerialName("needs_match")
    NEEDS_MATCH("needs_match", "Needs match"),

    /** Sources disagree. Never silently resolved — a human decides. */
    @SerialName("conflict")
    CONFLICT("conflict", "Verification conflict"),
    ;

    val detail: String
        get() = when (this) {
            VERIFIED -> "Product identity and activity groups confirmed against authoritative sources."
            PARTIALLY_VERIFIED -> "Product identified, but some resistance information is still unconfirmed."
            UNVERIFIED -> "Entered manually or carried over from an older record. Not confirmed against a label."
            NEEDS_MATCH -> "Not yet matched to a registered product for this country."
            CONFLICT -> "Sources disagree. Resolve before relying on this product's resistance information."
        }

    /**
     * Whether the future Resistance Engine may treat this product's groups as
     * dependable without warning the operator.
     */
    val isResistanceDependable: Boolean get() = this == VERIFIED

    /** Ranking used when merging: a merge may only ever LOWER confidence. */
    val confidenceRank: Int
        get() = when (this) {
            VERIFIED -> 4
            PARTIALLY_VERIFIED -> 3
            NEEDS_MATCH -> 2
            UNVERIFIED -> 1
            CONFLICT -> 0
        }

    companion object {
        /** An unknown status degrades to unverified — the only safe direction. */
        fun from(raw: String?): ChemicalVerificationStatus {
            val v = raw?.trim()?.lowercase()?.takeIf { it.isNotEmpty() } ?: return UNVERIFIED
            return entries.firstOrNull { it.raw == v } ?: UNVERIFIED
        }
    }
}

/**
 * A specific disagreement between two sources about one field.
 *
 * Surfaced to the operator verbatim — VineTrack does not pick a winner, and it
 * certainly does not quietly keep the AI's answer.
 */
@Serializable
data class ChemicalVerificationConflict(
    /** Which field disagrees, e.g. `"activity_group"`, `"concentration"`. */
    val field: String = "",
    @SerialName("active_ingredient_name") val activeIngredientName: String? = null,
    /** What the label/AI extraction claimed. */
    @SerialName("extracted_value") val extractedValue: String = "",
    /** What the authoritative classification says. */
    @SerialName("authoritative_value") val authoritativeValue: String = "",
    @SerialName("extracted_source")
    val extractedSource: ChemicalDataSourceKind = ChemicalDataSourceKind.AI_INTERPRETATION,
    @SerialName("authoritative_source")
    val authoritativeSource: ChemicalDataSourceKind = ChemicalDataSourceKind.AUTHORITATIVE_CLASSIFICATION,
) {
    // `this.field` is deliberate: inside a property accessor a bare `field` is
    // the backing-field keyword, not this class's `field` property.
    val id: String
        get() = "${this.field}|${activeIngredientName.orEmpty()}|$extractedValue|$authoritativeValue"

    /** One-line operator-facing summary. */
    val summary: String
        get() {
            val subject = activeIngredientName?.let { "$it: " } ?: ""
            return "$subject${extractedSource.label} says $extractedValue; " +
                "${authoritativeSource.label} says $authoritativeValue."
        }
}

/** The full verification picture for one chemical. */
@Serializable
data class ChemicalVerification(
    /**
     * The stored status. Prefer [resolvedStatus] when deciding what to trust —
     * it re-derives the claim from the evidence rather than believing a status
     * somebody wrote down.
     */
    val status: ChemicalVerificationStatus = ChemicalVerificationStatus.UNVERIFIED,
    val sources: List<ChemicalDataSource> = emptyList(),
    @SerialName("verified_at") val verifiedAt: String? = null,
    /** Unresolved disagreements. Non-empty forces [ChemicalVerificationStatus.CONFLICT]. */
    val conflicts: List<ChemicalVerificationConflict> = emptyList(),
    /**
     * Fields the lookup explicitly could not resolve, named so the UI can show
     * what is missing instead of an unexplained blank.
     */
    @SerialName("unresolved_fields") val unresolvedFields: List<String> = emptyList(),
) {
    /**
     * The status the EVIDENCE supports, which may be lower than [status].
     *
     * This is the gate that makes source-disagreement handling structural rather
     * than advisory: a record can carry `VERIFIED`, but if a conflict is present
     * or an active's group is unconfirmed, this returns `CONFLICT` or
     * `PARTIALLY_VERIFIED` and the product is not treated as dependable.
     * Confidence can only ever be lowered here, never raised.
     */
    fun resolvedStatus(
        actives: List<ChemicalActiveIngredient>,
        hasRegistration: Boolean,
    ): ChemicalVerificationStatus {
        if (conflicts.isNotEmpty()) return ChemicalVerificationStatus.CONFLICT
        if (actives.isEmpty()) {
            return if (status == ChemicalVerificationStatus.NEEDS_MATCH) {
                ChemicalVerificationStatus.NEEDS_MATCH
            } else {
                ChemicalVerificationStatus.UNVERIFIED
            }
        }
        val everyGroupAuthoritative = actives.all { it.hasAuthoritativeGroup }

        if (status == ChemicalVerificationStatus.VERIFIED &&
            everyGroupAuthoritative &&
            hasRegistration &&
            sources.containsAuthoritative() &&
            unresolvedFields.isEmpty()
        ) {
            return ChemicalVerificationStatus.VERIFIED
        }

        // Partially verified means "we know WHICH product this is, but something
        // about it is still unconfirmed". It therefore requires real evidence:
        // an authoritative registered identity, an authoritative cited source, or
        // an active whose group an authoritative classification stands behind.
        //
        // Merely HAVING a group is not enough. A chemical someone typed by hand
        // has a group because they typed one, and that must stay Unverified —
        // otherwise manual entry quietly launders itself into partial trust.
        val hasAuthoritativeEvidence = hasRegistration ||
            sources.containsAuthoritative() ||
            actives.any { it.hasAuthoritativeGroup }

        if (hasAuthoritativeEvidence) {
            return if (status == ChemicalVerificationStatus.NEEDS_MATCH) {
                ChemicalVerificationStatus.NEEDS_MATCH
            } else {
                ChemicalVerificationStatus.PARTIALLY_VERIFIED
            }
        }
        return if (status == ChemicalVerificationStatus.NEEDS_MATCH) {
            ChemicalVerificationStatus.NEEDS_MATCH
        } else {
            ChemicalVerificationStatus.UNVERIFIED
        }
    }

    /** Records a disagreement and drops the status to conflict. */
    fun addingConflict(conflict: ChemicalVerificationConflict): ChemicalVerification =
        if (conflicts.contains(conflict)) this
        else copy(
            conflicts = conflicts + conflict,
            status = ChemicalVerificationStatus.CONFLICT,
        )

    companion object {
        /** Verification state for a chemical the operator typed in themselves. */
        fun manual(): ChemicalVerification = ChemicalVerification(
            status = ChemicalVerificationStatus.UNVERIFIED,
            sources = listOf(
                ChemicalDataSource(
                    kind = ChemicalDataSourceKind.MANUAL_ENTRY,
                    name = "Entered in VineTrack",
                ),
            ),
        )

        /**
         * Verification state for a pre-Chemical-Intelligence record. It is not
         * "unverified because it's wrong" — it simply has never been matched to a
         * registered product, and the audit needs to tell those apart.
         */
        fun legacy(): ChemicalVerification = ChemicalVerification(
            status = ChemicalVerificationStatus.NEEDS_MATCH,
            sources = listOf(
                ChemicalDataSource(
                    kind = ChemicalDataSourceKind.LEGACY_RECORD,
                    name = "Existing VineTrack chemical",
                ),
            ),
        )
    }
}

File: ios/VineTrack/App/ChemicalIntelligence/ChemicalVerification.swift
import Foundation

/// How much VineTrack trusts a chemical's resistance-critical information.
///
/// This is deliberately separate from "did the lookup return anything". An AI
/// answer is a lead, not a verification: a product only becomes `.verified`
/// when an authoritative source stands behind its identity AND behind every
/// active's activity group.
nonisolated enum ChemicalVerificationStatus: String, Codable, Sendable, CaseIterable, Hashable {
    /// Product identity and every active's activity group are confirmed
    /// against authoritative sources.
    case verified
    /// Product identity is established, but at least one resistance-relevant
    /// field is still unconfirmed.
    case partiallyVerified = "partially_verified"
    /// Operator-entered or legacy data, or too little evidence to classify.
    case unverified
    /// A legacy record that has never been put through the match step. It has
    /// data, but nobody has yet confirmed WHICH registered product it is.
    case needsMatch = "needs_match"
    /// Sources disagree about a resistance-critical field. Never silently
    /// resolved — a human decides.
    case conflict

    nonisolated var label: String {
        switch self {
        case .verified: return "Verified"
        case .partiallyVerified: return "Partially verified"
        case .unverified: return "Unverified"
        case .needsMatch: return "Needs match"
        case .conflict: return "Verification conflict"
        }
    }

    nonisolated var detail: String {
        switch self {
        case .verified:
            return "Product identity and activity groups confirmed against authoritative sources."
        case .partiallyVerified:
            return "Product identified, but some resistance information is still unconfirmed."
        case .unverified:
            return "Entered manually or carried over from an older record. Not confirmed against a label."
        case .needsMatch:
            return "Not yet matched to a registered product for this country."
        case .conflict:
            return "Sources disagree. Resolve before relying on this product's resistance information."
        }
    }

    /// Whether the future Resistance Engine may treat this product's groups as
    /// dependable without warning the operator.
    nonisolated var isResistanceDependable: Bool { self == .verified }

    /// Ranking used when merging: a merge may only ever LOWER confidence.
    nonisolated var confidenceRank: Int {
        switch self {
        case .verified: return 4
        case .partiallyVerified: return 3
        case .needsMatch: return 2
        case .unverified: return 1
        case .conflict: return 0
        }
    }
}

/// A specific disagreement between two sources about one field.
///
/// Surfaced to the operator verbatim — VineTrack does not pick a winner, and
/// it certainly does not quietly keep the AI's answer.
nonisolated struct ChemicalVerificationConflict: Codable, Sendable, Hashable, Identifiable {
    /// Which field disagrees, e.g. `"activity_group"`, `"concentration"`.
    let field: String
    /// The active ingredient the disagreement concerns, when field-specific.
    let activeIngredientName: String?
    /// What the label/AI extraction claimed.
    let extractedValue: String
    /// What the authoritative classification says.
    let authoritativeValue: String
    /// Source that produced `extractedValue`.
    let extractedSource: ChemicalDataSourceKind
    /// Source that produced `authoritativeValue`.
    let authoritativeSource: ChemicalDataSourceKind

    nonisolated var id: String {
        "\(field)|\(activeIngredientName ?? "")|\(extractedValue)|\(authoritativeValue)"
    }

    init(
        field: String,
        activeIngredientName: String? = nil,
        extractedValue: String,
        authoritativeValue: String,
        extractedSource: ChemicalDataSourceKind = .aiInterpretation,
        authoritativeSource: ChemicalDataSourceKind = .authoritativeClassification
    ) {
        self.field = field
        self.activeIngredientName = activeIngredientName
        self.extractedValue = extractedValue
        self.authoritativeValue = authoritativeValue
        self.extractedSource = extractedSource
        self.authoritativeSource = authoritativeSource
    }

    nonisolated enum CodingKeys: String, CodingKey {
        case field
        case activeIngredientName = "active_ingredient_name"
        case extractedValue = "extracted_value"
        case authoritativeValue = "authoritative_value"
        case extractedSource = "extracted_source"
        case authoritativeSource = "authoritative_source"
    }

    nonisolated init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        field = try c.decodeIfPresent(String.self, forKey: .field) ?? ""
        activeIngredientName = try c.decodeIfPresent(String.self, forKey: .activeIngredientName)
        extractedValue = try c.decodeIfPresent(String.self, forKey: .extractedValue) ?? ""
        authoritativeValue = try c.decodeIfPresent(String.self, forKey: .authoritativeValue) ?? ""
        let ex = try c.decodeIfPresent(String.self, forKey: .extractedSource) ?? ""
        extractedSource = ChemicalDataSourceKind(rawValue: ex) ?? .aiInterpretation
        let au = try c.decodeIfPresent(String.self, forKey: .authoritativeSource) ?? ""
        authoritativeSource = ChemicalDataSourceKind(rawValue: au) ?? .authoritativeClassification
    }

    /// One-line operator-facing summary.
    nonisolated var summary: String {
        let subject = activeIngredientName.map { "\($0): " } ?? ""
        return "\(subject)\(extractedSource.label) says \(extractedValue); "
            + "\(authoritativeSource.label) says \(authoritativeValue)."
    }
}

/// The full verification picture for one chemical.
nonisolated struct ChemicalVerification: Codable, Sendable, Hashable {
    /// The stored status. Prefer `resolvedStatus(actives:hasRegistration:)`
    /// when deciding what to trust — it re-derives the claim from the evidence
    /// rather than believing a status somebody wrote down.
    var status: ChemicalVerificationStatus
    /// Every source consulted.
    var sources: [ChemicalDataSource]
    /// When verification last ran.
    var verifiedAt: Date?
    /// Unresolved disagreements. Non-empty forces `.conflict`.
    var conflicts: [ChemicalVerificationConflict]
    /// Fields the lookup explicitly could not resolve, named so the UI can
    /// show what is missing instead of an unexplained blank.
    var unresolvedFields: [String]

    init(
        status: ChemicalVerificationStatus = .unverified,
        sources: [ChemicalDataSource] = [],
        verifiedAt: Date? = nil,
        conflicts: [ChemicalVerificationConflict] = [],
        unresolvedFields: [String] = []
    ) {
        self.status = status
        self.sources = sources
        self.verifiedAt = verifiedAt
        self.conflicts = conflicts
        self.unresolvedFields = unresolvedFields
    }

    nonisolated enum CodingKeys: String, CodingKey {
        case status, sources, conflicts
        case verifiedAt = "verified_at"
        case unresolvedFields = "unresolved_fields"
    }

    nonisolated init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let raw = try c.decodeIfPresent(String.self, forKey: .status) ?? ""
        // An unknown status from a newer build degrades to unverified. Erring
        // downward is the only safe direction for a trust claim.
        status = ChemicalVerificationStatus(rawValue: raw) ?? .unverified
        sources = try c.decodeIfPresent([ChemicalDataSource].self, forKey: .sources) ?? []
        // ISO-8601 string on the wire (edge function/portal), Date under a
        // strategy-configured decoder, or null — all must decode.
        verifiedAt = ChemicalWireDate.decode(from: c, key: .verifiedAt)
        conflicts = try c.decodeIfPresent([ChemicalVerificationConflict].self, forKey: .conflicts) ?? []
        unresolvedFields = try c.decodeIfPresent([String].self, forKey: .unresolvedFields) ?? []
    }

    /// The status the EVIDENCE supports, which may be lower than `status`.
    ///
    /// This is the gate that makes Phase 7 structural rather than advisory: a
    /// record can carry `status == .verified`, but if a conflict is present or
    /// an active's group is unconfirmed, this returns `.conflict` or
    /// `.partiallyVerified` and the product is not treated as dependable.
    /// Confidence can only ever be lowered here, never raised.
    func resolvedStatus(
        actives: [ChemicalActiveIngredient],
        hasRegistration: Bool
    ) -> ChemicalVerificationStatus {
        if !conflicts.isEmpty { return .conflict }
        if actives.isEmpty {
            return status == .needsMatch ? .needsMatch : .unverified
        }
        let everyGroupAuthoritative = actives.allSatisfy(\.hasAuthoritativeGroup)

        if status == .verified,
           everyGroupAuthoritative,
           hasRegistration,
           sources.containsAuthoritative,
           unresolvedFields.isEmpty {
            return .verified
        }

        // Partially verified means "we know WHICH product this is, but something
        // about it is still unconfirmed". It therefore requires real evidence:
        // an authoritative registered identity, an authoritative cited source, or
        // an active whose group an authoritative classification stands behind.
        //
        // Merely HAVING a group is not enough. A chemical someone typed by hand
        // has a group because they typed one, and that must stay Unverified —
        // otherwise manual entry quietly launders itself into partial trust.
        let hasAuthoritativeEvidence = hasRegistration
            || sources.containsAuthoritative
            || actives.contains(where: \.hasAuthoritativeGroup)

        if hasAuthoritativeEvidence {
            return status == .needsMatch ? .needsMatch : .partiallyVerified
        }
        return status == .needsMatch ? .needsMatch : .unverified
    }

    /// Records a disagreement and drops the status to `.conflict`.
    mutating func addConflict(_ conflict: ChemicalVerificationConflict) {
        guard !conflicts.contains(conflict) else { return }
        conflicts.append(conflict)
        status = .conflict
    }

    /// Verification state for a chemical the operator typed in themselves.
    static func manual() -> ChemicalVerification {
        ChemicalVerification(
            status: .unverified,
            sources: [ChemicalDataSource(kind: .manualEntry, name: "Entered in VineTrack")],
            verifiedAt: nil
        )
    }

    /// Verification state for a pre-Chemical-Intelligence record. It is not
    /// "unverified because it's wrong" — it simply has never been matched to a
    /// registered product, and the audit needs to be able to tell those apart.
    static func legacy() -> ChemicalVerification {
        ChemicalVerification(
            status: .needsMatch,
            sources: [ChemicalDataSource(kind: .legacyRecord, name: "Existing VineTrack chemical")],
            verifiedAt: nil
        )
    }
}
